library/core/src/ptr/mod.rs

```diff
@@ -65,6 +65,8 @@
 //! separate allocated object), heap allocations (each allocation created by the global allocator is
 //! a separate allocated object), and `static` variables.
```

Comment:
Should we maybe extend the part above a bit, taking some wording from the [UCG](https://rust-lang.github.io/unsafe-code-guidelines/glossary.html#allocation)?
Like "An allocation object is a chunk of memory that can be accessed from Rust, and within which pointer arithmetic can be used to offset a pointer from one part of the allocated object to another.  Allocated objects are currently always contiguous and their size never changes between allocation and deallocation, but in the future, we might allow allocated objects with holes, and we might allow growing/shrinking an allocated object. Common examples ..."

Reply:
How about the following? I tried to only keep the bits that are observable from code running in the Rust abstract machine (e.g., I removed the bit about functions and vtables). 

> An *allocated object* is a subset of program memory which is addressable from Rust. Allocated objects are created for objects on the heap, for stack-allocated variables, and for global statics and consts (note that this may not be an exhaustive list). The safety preconditions of some Rust operations - such as `offset` and field projections (`expr.field`) - are defined in terms of the allocated objects on which they operate.
>
> An allocated object has a base address and a set of memory addresses. It is possible for an allocated object to have no addresses (i.e., to be "empty"), but such an allocated object will still have a base address. The base address of an allocated object is not necessarily unique, but if two distinct allocated objects have the same base address then at least one of them must be empty. While it is currently the case that an allocated object always has a set of memory addresses which is fully contiguous (i.e., has no "holes"), there is no guarantee that this will not change in the future.

The glossary says that, in the future, "we might allow growing/shrinking an allocation." Is the idea here that an allocation's set of addresses could change over time so that observing an allocation to have a particular set of addresses does not necessarily guarantee that at a future point in the program execution it will still have that set of addresses? I'm wondering whether it's worth mentioning this and, if so, how to phrase it.

Also, from a prose perspective, it might be useful to also introduce "allocation" as a synonym. The second paragraph contains the phrase "allocated object" enough times that it's kind of awkward.

Reply:
> (i.e., to be "empty"),

I would say "zero-sized", that's the term people are more likely to encounter.

I know it's copied from the UCG but I am not super happy with just "a subset of program memory which is addressable from Rust". According to that definition, why isn't there just one big allocation that's the whole set of addressable memory? "One can do pointer arithmetic inside an allocation" seems to be a defining property. What about something like:

> An allocated object is a subset of program memory which is addressable from Rust, and within which pointer arithmetic is possible.

(we should also adjust the UCG glossary then)

> Also, from a prose perspective, it might be useful to also introduce "allocation" as a synonym. The second paragraph contains the phrase "allocated object" enough times that it's kind of awkward.

I think there is some concern that "allocation" might imply "heap allocation" (as in the `std::alloc` module and "global allocator" and so on).

Reply:
Okay, based on those edits:

> An *allocated object* is a subset of program memory which is addressable from Rust, and within which pointer arithmetic is possible. Examples of allocated objects include heap allocations, stack-allocated variables, statics, and consts. The safety preconditions of some Rust operations - such as `offset` and field projections (`expr.field`) - are defined in terms of the allocated objects on which they operate.
>
> An allocated object has a base address and a set of memory addresses. It is possible for an allocated object to have no addresses (i.e., to be "empty"), but such an allocated object will still have a base address. The base address of an allocated object is not necessarily unique, but if two distinct allocated objects have the same base address then at least one of them must be empty. While it is currently the case that an allocated object always has a set of memory addresses which is fully contiguous (i.e., has no "holes"), there is no guarantee that this will not change in the future.

Reply:
Yes, I like that.

Now, for the other part we also need allocated objects to have a size. Should we say that's the distance between the lowest and highest address contained in the object? That would then not assume things to be contiguous.

Reply:
Leaving a note to myself: The outstanding edits are:

> Now, for the other part we also need allocated objects to have a size. Should we say that's the distance between the lowest and highest address contained in the object? That would then not assume things to be contiguous.

...and from https://github.com/rust-lang/rust/pull/116675#discussion_r1368287834 (which is based on the current text in the PR rather than the draft described in this thread):

> I see only one potential problem: if the object has size zero, "one byte past the last address in the object" is ill-defined as there is no address in the object. I never really like this "one past the end" terminology either. This is just "base_addr + size", and I think of it as the pointer "at the end".

Reply:
New draft:

> An *allocated object* is a subset of program memory which is addressable from Rust, and within which pointer arithmetic is possible. Examples of allocated objects include heap allocations, stack-allocated variables, statics, and consts. The safety preconditions of some Rust operations - such as `offset` and field projections (`expr.field`) - are defined in terms of the allocated objects on which they operate.
>
> An allocated object has a base address and a set of memory addresses. It is possible for an allocated object to have no addresses (i.e., to be "zero-sized"), but such an allocated object will still have a base address. The base address of an allocated object is not necessarily unique, but if two distinct allocated objects have the same base address then at least one of them must be zero-sized. While it is currently the case that an allocated object always has a set of memory addresses which is fully contiguous (i.e., has no "holes"), there is no guarantee that this will not change in the future.
>
> Assume that an allocated object has a base address, `base`. Let `end` be defined as:
> - If the allocated object contains no addresses, `end` is `base + 1`
> - If the allocated object contains at least one address, and the numerically greatest address it contains has the value `n`, then `end` is `n + 1`
>
> The following are guaranteed about any such allocated object:
> - `base` is not equal to `null()` (i.e., the address with the numerical value 0)
> - `base` is not equal to `usize::MAX`
> - For all addresses, `a`, contained in the allocated object:
>   - `base <= a`
>   - `a < usize::MAX`
>   - `(a + 1) - base <= isize::MAX`
>
> As a consequence of these guarantees, given the offset, `o`, from `base` of any address, `a`, which is either in the allocated object or is the address `end`:
> - It is guaranteed that `o` does not overflow `isize`
> - It is guaranteed that `o` is non-negative
> - It is possible to compute the `a` as `base + o`, and this addition will not wrap around the address space (in other words, will not overflow `usize`)

Note a subtlety here: The current PR text includes:

> An allocated object ... cannot contain the last (`usize::MAX`) byte of the address space. As a consequence, any address which is either in the allocated object or one byte past the last address in the allocated object can be computed without wrapping around the address space.

If we don't guarantee that the set of addresses contained in an allocated object is contiguous, then the second sentence is not actually a logical implication of the first sentence. I believe that the updated draft does not have this logic hole.

Reply:
Thanks for the proposal! I largely like it.

However, I don't like the fact that this has to special-case for size zero. I think if I had to write this up myself, I'd say that an allocated object has a base address `base` and a size `size`, such that all addresses that belong to the allocated object are contained in `base .. (base+size)`. For now, all allocated objects are contiguous, and `base .. (base+size)` are *exactly* the set of addresses that belong to the allocated object; but in the future, we might support allocations with "holes" inside of them. We have that
- `base > 0`
- `size <= isize::MAX`
- `base + size <= usize::MAX` (using infinite-precision arithmetic for the `+`)

Do we actually need to say that base addresses are unique? There are some unfortunate issues around pointer identity and some of the proposals for resolving them involve letting the abstract machine have "overlapping" allocations that are disambiguated by provenance. (When the program runs on the real hardware, of course base addresses are unique, but some allocations might be optimized away and the only way to explain how that all happened is that in the AM execution, the allocations overlapped.)

> As a consequence of these guarantees, given the offset, o, from base of any address, a, which is either in the allocated object or is the address end:

I can't parse the "from base of any address" part.

Reply:
> However, I don't like the fact that this has to special-case for size zero. I think if I had to write this up myself, I'd say that an allocated object has a base address `base` and a size `size`, such that all addresses that belong to the allocated object are contained in `base .. (base+size)`. For now, all allocated objects are contiguous, and `base .. (base+size)` are _exactly_ the set of addresses that belong to the allocated object; but in the future, we might support allocations with "holes" inside of them. We have that
> 
> * `base > 0`
> * `size <= isize::MAX`
> * `base + size <= usize::MAX` (using infinite-precision arithmetic for the `+`)

This sounds good! I'll update the draft accordingly.

> Do we actually need to say that base addresses are unique? There are some unfortunate issues around pointer identity and some of the proposals for resolving them involve letting the abstract machine have "overlapping" allocations that are disambiguated by provenance. (When the program runs on the real hardware, of course base addresses are unique, but some allocations might be optimized away and the only way to explain how that all happened is that in the AM execution, the allocations overlapped.)

I suppose we could just remove the uniqueness language and we just wouldn't be providing as strong of a guarantee. Alternatively, maybe we could leave the uniqueness language in and say "today" and then provide the standard "no guarantee this won't change" caveat? That approach has the advantage that it's easier to update this text in the future if we do commit to such a guarantee - we just remove the caveat language.

> > As a consequence of these guarantees, given the offset, o, from base of any address, a, which is either in the allocated object or is the address end:
> 
> I can't parse the "from base of any address" part.

Bad wording on my part. This is supposed to be read as:

> As a consequence of these guarantees, given the offset (called `o`, evaluated relative to `base`) of any address (called `a`) which is either in the allocated object or is the address `end`:

I'll update the wording to be more clear.

---

<details>
<summary>New draft with all of these edits</summary>

> An *allocated object* is a subset of program memory which is addressable from Rust, and within which pointer arithmetic is possible. Examples of allocated objects include heap allocations, stack-allocated variables, statics, and consts. The safety preconditions of some Rust operations - such as `offset` and field projections (`expr.field`) - are defined in terms of the allocated objects on which they operate.
>
> An allocated object has a base address, a size, and a set of memory addresses. It is possible for an allocated object to have zero size, but such an allocated object will still have a base address. The base address of an allocated object is not necessarily unique. The following two facts are currently true in Rust, but no guarantee is provided that they will not change in the future:
> - If two distinct allocated objects have the same base address, then at least one of them must be zero-sized.
> - An allocated object always has a set of memory addresses which is fully contiguous (i.e., has no "holes").
>
> For any allocated object with `base` address, `size`, and a set of `addresses`, the following are guaranteed:
> - For all addresses, `a`, in `addresses`, `base <= a < base + size`
> - `base` is not equal to `null()` (i.e., the address with the numerical value 0)
> - `base` is not equal to `usize::MAX`
> - `base + size <= usize::MAX`
> - `size <= isize::MAX`
>
> As a consequence of these guarantees, given any address, `a`, within the set of addresses of an allocated object:
> - It is guaranteed that `a - base` does not overflow `isize`
> - It is guaranteed that `a - base` is non-negative
> - It is guaranteed that, given `o = a - base` (i.e., the offset of `a` within the allocated object), `base + o` will not wrap around the address space (in other words, will not overflow `usize`)

</details>

Reply:
> I suppose we could just remove the uniqueness language and we just wouldn't be providing as strong of a guarantee. Alternatively, maybe we could leave the uniqueness language in and say "today" and then provide the standard "no guarantee this won't change" caveat? That approach has the advantage that it's easier to update this text in the future if we do commit to such a guarantee - we just remove the caveat language.

That seems worse than just saying nothing, honestly.

> base is not equal to usize::MAX

Is this true? I think for zero-sized allocations we'd be in our right to put them at usize::MAX.

Reply:
> > I suppose we could just remove the uniqueness language and we just wouldn't be providing as strong of a guarantee. Alternatively, maybe we could leave the uniqueness language in and say "today" and then provide the standard "no guarantee this won't change" caveat? That approach has the advantage that it's easier to update this text in the future if we do commit to such a guarantee - we just remove the caveat language.
> 
> That seems worse than just saying nothing, honestly.

Sounds good; I'll remove it.

> > base is not equal to usize::MAX
> 
> Is this true? I think for zero-sized allocations we'd be in our right to put them at usize::MAX.
> 
> "base + size <= usize::MAX" is the key point.

Yeah I suppose you're right. I've removed that requirement in the new draft below.

I've also reworded this:

> - For all addresses, `a`, in addresses, `base <= a < base + size`

This is badly worded since it implies that `base < base + size`, ie that `size > 0`; this is actually okay since it only holds for allocated objects which have at least one address so that the "forall" is actually inhabited, but it's needlessly confusing.


---

<details>
<summary>New draft with all of these edits</summary>

> An *allocated object* is a subset of program memory which is addressable from Rust, and within which pointer arithmetic is possible. Examples of allocated objects include heap allocations, stack-allocated variables, statics, and consts. The safety preconditions of some Rust operations - such as `offset` and field projections (`expr.field`) - are defined in terms of the allocated objects on which they operate.
>
> An allocated object has a base address, a size, and a set of memory addresses. It is possible for an allocated object to have zero size, but such an allocated object will still have a base address. The base address of an allocated object is not necessarily unique. While it is currently the case that an allocated object always has a set of memory addresses which is fully contiguous (i.e., has no "holes"), there is no guarantee that this will not change in the future.
>
> For any allocated object with `base` address, `size`, and a set of `addresses`, the following are guaranteed:
> - For all addresses, `a`, in `addresses`, `a` is in the range `[base, base + size)`
> - `base` is not equal to `null()` (i.e., the address with the numerical value 0)
> - `base + size <= usize::MAX`
> - `size <= isize::MAX`
>
> As a consequence of these guarantees, given any address, `a`, within the set of addresses of an allocated object:
> - It is guaranteed that `a - base` does not overflow `isize`
> - It is guaranteed that `a - base` is non-negative
> - It is guaranteed that, given `o = a - base` (i.e., the offset of `a` within the allocated object), `base + o` will not wrap around the address space (in other words, will not overflow `usize`)

</details>

Reply:
> This is badly worded since it implies that base < base + size, ie that size > 0; this is actually okay since it only holds for allocated objects which have at least one address so that the "forall" is actually inhabited, but it's needlessly confusing.

I didn't find it confusing. I'm also fine with using a range instead, but then I'd suggest using Rust range syntax: `base .. (base+size)`.

Reply:
> I didn't find it confusing. I'm also fine with using a range instead, but then I'd suggest using Rust range syntax: `base .. (base+size)`.

IMO the `[x, y)` syntax is clearer because the fact that `[` and `)` are different character types (as opposed to `[x, y]` or `(x, y)`) calls your attention to the discrepancy. I worry it'd be easy to forget that the Rust range syntax refers to a closed/open range, whereas, reading `[x, y)`, you're unlikely not to notice that it's a closed/open range.

Reply:
I quite dislike the usual math notation because these unbalanced parentheses just hurt my eyes. ;)

More seriously though, I think it is also far from self-explaining syntax. We have syntax that we can expect the reader of this to be familiar with; we should use that syntax.

Reply:
Yeah that's a fair point about consistency (wake me up when [rustdoc supports inline LaTeX](https://github.com/rust-lang/rust/issues/34261)...). How about:

> - For all addresses, `a`, in `addresses`, `a` is in the range `base .. (base + size)` (note that this requires `a < base + size`, not `a <= base + size`)

Reply:
Seems unnecessary to me to explain the meaning of our default `..` ranges, but :shrug: no strong opinion on that point. Fine otherwise.

Reply:
Btw why do you have commas around the first `a`? I would have written

> For all addresses `a` in `addresses`, `a` is in the range ...

Reply:
> Seems unnecessary to me to explain the meaning of our default `..` ranges, but 🤷 no strong opinion on that point. Fine otherwise.

I grant that it won't be common for this clarification to be necessary, but I'd prefer to err on the side of caution given how important the semantics are here.

> Btw why do you have commas around the first `a`? I would have written
> 
> > For all addresses `a` in `addresses`, `a` is in the range ...

I'm thinking in terms of when you introduce names, e.g.: "My friend, Bob, said..." But I don't mind the syntax you suggested; I've updated it here:

<details>
<summary>New draft with both of these edits</summary>

> An *allocated object* is a subset of program memory which is addressable from Rust, and within which pointer arithmetic is possible. Examples of allocated objects include heap allocations, stack-allocated variables, statics, and consts. The safety preconditions of some Rust operations - such as `offset` and field projections (`expr.field`) - are defined in terms of the allocated objects on which they operate.
>
> An allocated object has a base address, a size, and a set of memory addresses. It is possible for an allocated object to have zero size, but such an allocated object will still have a base address. The base address of an allocated object is not necessarily unique. While it is currently the case that an allocated object always has a set of memory addresses which is fully contiguous (i.e., has no "holes"), there is no guarantee that this will not change in the future.
>
> For any allocated object with `base` address, `size`, and a set of `addresses`, the following are guaranteed:
> - For all addresses `a` in `addresses`, `a` is in the range `base .. (base + size)` (note that this requires `a < base + size`, not `a <= base + size`)
> - `base` is not equal to `null()` (i.e., the address with the numerical value 0)
> - `base + size <= usize::MAX`
> - `size <= isize::MAX`
>
> As a consequence of these guarantees, given any address, `a`, within the set of addresses of an allocated object:
> - It is guaranteed that `a - base` does not overflow `isize`
> - It is guaranteed that `a - base` is non-negative
> - It is guaranteed that, given `o = a - base` (i.e., the offset of `a` within the allocated object), `base + o` will not wrap around the address space (in other words, will not overflow `usize`)

</details>